src/App.jsx

```diff
@@ -14,33 +15,110 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
-export const App = () => (
-  <div className="section content">
-    <div className="buttons">
-      <button type="button" className="button is-info is-light">
-        Sort alphabetically
-      </button>
+const prepareGoods = (goods, { field, isReverse }) => {
+  const sortedGoods = [...goods];
 
-      <button type="button" className="button is-success is-light">
-        Sort by length
-      </button>
+  if (field === 'alphabetical') {
+    sortedGoods.sort();
+  } else if (field === 'length') {
+    sortedGoods.sort((a, b) => a.length - b.length);
+  }
 
-      <button type="button" className="button is-warning is-light">
-        Reverse
-      </button>
+  if (isReverse) {
+    sortedGoods.reverse();
+  }
 
-      <button type="button" className="button is-danger is-light">
-        Reset
-      </button>
+  return sortedGoods;
+};
+
+export const App = () => {
+  const [visibleGoods, setVisibleGoods] = useState(goodsFromServer);
+  const [isReversed, setIsReversed] = useState(false);
+  const [sortField, setSortField] = useState(null);
+  const [originalGoods] = useState(goodsFromServer);
+
+  const handleSortByName = () => {
+    setSortField('alphabetical');
+    setVisibleGoods(
+      prepareGoods(goodsFromServer, {
+        field: 'alphabetical',
+        isReverse: isReversed,
+      }),
+    );
+  };
+
+  const handleSortByLength = () => {
+    setSortField('length');
+    setVisibleGoods(
+      prepareGoods(goodsFromServer, { field: 'length', isReverse: isReversed }),
+    );
+  };
+
+  const handleReverse = () => {
+    setIsReversed(!isReversed);
+    setVisibleGoods(
+      prepareGoods(goodsFromServer, {
+        field: sortField,
+        isReverse: !isReversed,
+      }),
+    );
+  };
+
+  const handleReset = () => {
+    setVisibleGoods(originalGoods);
+    setSortField(null);
+    setIsReversed(false);
+  };
+
+  const showResetButton = visibleGoods.join() !== originalGoods.join();
```

Comment:
The `showResetButton` logic should ensure that the reset button is only shown when the visible goods are not in the original order. The current implementation uses `join()` to compare arrays, which may not be reliable for complex data structures.